From 8073ac66c76236eb60ef900d81650de690a1e3b6 Mon Sep 17 00:00:00 2001
From: Mirko Damiani <mirko@develer.com>
Date: Thu, 10 Dec 2015 12:58:51 +0100
Subject: [PATCH] add support for ddr2 sdram at 200MHz
Organization: Develer S.r.l.

ddr2 sdram list:
 - Samsung K4T51163QQ-BIE6
 - Samsung K4T1G164QG-BCE6
 - Micron MT47H128M16RT-25E
---
 board/sama5d4_xplained/sama5d4_xplained.c | 117 +++---------------------------
 include/common.h                          |   8 +-
 main.c                                    | 112 ++++++++++++++++++++++++++++
 3 files changed, 130 insertions(+), 107 deletions(-)

diff --git a/board/sama5d4_xplained/sama5d4_xplained.c b/board/sama5d4_xplained/sama5d4_xplained.c
index 47eb111..82bb9a4 100644
--- a/board/sama5d4_xplained/sama5d4_xplained.c
+++ b/board/sama5d4_xplained/sama5d4_xplained.c
@@ -95,120 +95,25 @@ static void initialize_dbgu(void)
 #ifdef CONFIG_DDR2
 static void ddramc_reg_config(struct ddramc_register *ddramc_config)
 {
-	ddramc_config->mdr = (AT91C_DDRC2_DBW_32_BITS
-				| AT91C_DDRC2_MD_DDR2_SDRAM);
+	ddramc_config->mdr = _at91_parameters.ddramc_mdr;
+	ddramc_config->cr = _at91_parameters.ddramc_cr;
 
-	ddramc_config->cr = (AT91C_DDRC2_NC_DDR10_SDR9
-				| AT91C_DDRC2_NR_14
-				| AT91C_DDRC2_CAS_3
-				| AT91C_DDRC2_DLL_RESET_DISABLED
-				| AT91C_DDRC2_DIS_DLL_DISABLED
-				| AT91C_DDRC2_NB_BANKS_8
-				| AT91C_DDRC2_DECOD_INTERLEAVED
-				| AT91C_DDRC2_UNAL_SUPPORTED);
-
-
-#if defined(CONFIG_BUS_SPEED_148MHZ)
-
-	ddramc_config->rtr = 0x243;
-
-	/* One clock cycle @ 148 MHz = 6.7 ns */
-	ddramc_config->t0pr = (AT91C_DDRC2_TRAS_(7)
-			| AT91C_DDRC2_TRCD_(3)
-			| AT91C_DDRC2_TWR_(3)
-			| AT91C_DDRC2_TRC_(9)
-			| AT91C_DDRC2_TRP_(3)
-			| AT91C_DDRC2_TRRD_(2)
-			| AT91C_DDRC2_TWTR_(2)
-			| AT91C_DDRC2_TMRD_(2));
-
-	ddramc_config->t1pr = (AT91C_DDRC2_TXP_(2)
-			| AT91C_DDRC2_TXSRD_(200)
-			| AT91C_DDRC2_TXSNR_(31)
-			| AT91C_DDRC2_TRFC_(30));
-
-	ddramc_config->t2pr = (AT91C_DDRC2_TFAW_(7)
-			| AT91C_DDRC2_TRTP_(2)
-			| AT91C_DDRC2_TRPA_(3)
-			| AT91C_DDRC2_TXARDS_(8)
-			| AT91C_DDRC2_TXARD_(8));
-
-#elif defined(CONFIG_BUS_SPEED_170MHZ)
-
-	ddramc_config->rtr = 0x229;
-
-	/* One clock cycle @ 170 MHz = 5.9 ns */
-	ddramc_config->t0pr = (AT91C_DDRC2_TRAS_(7)
-			| AT91C_DDRC2_TRCD_(3)
-			| AT91C_DDRC2_TWR_(3)
-			| AT91C_DDRC2_TRC_(10)
-			| AT91C_DDRC2_TRP_(3)
-			| AT91C_DDRC2_TRRD_(2)
-			| AT91C_DDRC2_TWTR_(2)
-			| AT91C_DDRC2_TMRD_(2));
-
-	ddramc_config->t1pr = (AT91C_DDRC2_TXP_(2)
-			| AT91C_DDRC2_TXSRD_(200)
-			| AT91C_DDRC2_TXSNR_(24)
-			| AT91C_DDRC2_TRFC_(22));
-
-	ddramc_config->t2pr = (AT91C_DDRC2_TFAW_(6)
-			| AT91C_DDRC2_TRTP_(2)
-			| AT91C_DDRC2_TRPA_(3)
-			| AT91C_DDRC2_TXARDS_(2)
-			| AT91C_DDRC2_TXARD_(8));
-
-#elif defined(CONFIG_BUS_SPEED_176MHZ)
-
-	ddramc_config->rtr = 0x2b0;
-
-	ddramc_config->t0pr = (AT91C_DDRC2_TRAS_(8)
-			| AT91C_DDRC2_TRCD_(3)
-			| AT91C_DDRC2_TWR_(3)
-			| AT91C_DDRC2_TRC_(10)
-			| AT91C_DDRC2_TRP_(3)
-			| AT91C_DDRC2_TRRD_(2)
-			| AT91C_DDRC2_TWTR_(2)
-			| AT91C_DDRC2_TMRD_(2));
-
-	ddramc_config->t1pr = (AT91C_DDRC2_TXP_(2)
-			| AT91C_DDRC2_TXSRD_(200)
-			| AT91C_DDRC2_TXSNR_(25)
-			| AT91C_DDRC2_TRFC_(23));
-
-	ddramc_config->t2pr = (AT91C_DDRC2_TFAW_(7)
-			| AT91C_DDRC2_TRTP_(2)
-			| AT91C_DDRC2_TRPA_(3)
-			| AT91C_DDRC2_TXARDS_(2)
-			| AT91C_DDRC2_TXARD_(8));
-
-#elif defined(CONFIG_BUS_SPEED_200MHZ)
+#if defined(CONFIG_BUS_SPEED_200MHZ)
 
 	ddramc_config->rtr = 0x30e;
 
-	ddramc_config->t0pr = (AT91C_DDRC2_TRAS_(8)
-			| AT91C_DDRC2_TRCD_(3)
-			| AT91C_DDRC2_TWR_(3)
-			| AT91C_DDRC2_TRC_(11)
-			| AT91C_DDRC2_TRP_(3)
-			| AT91C_DDRC2_TRRD_(2)
-			| AT91C_DDRC2_TWTR_(2)
-			| AT91C_DDRC2_TMRD_(2));
-
-	ddramc_config->t1pr = (AT91C_DDRC2_TXP_(2)
-			| AT91C_DDRC2_TXSRD_(200)
-			| AT91C_DDRC2_TXSNR_(28)
-			| AT91C_DDRC2_TRFC_(26));
-
-	ddramc_config->t2pr = (AT91C_DDRC2_TFAW_(7)
-			| AT91C_DDRC2_TRTP_(2)
-			| AT91C_DDRC2_TRPA_(3)
-			| AT91C_DDRC2_TXARDS_(2)
-			| AT91C_DDRC2_TXARD_(8));
+	/* One clock cycle @ 200 MHz = 5 ns */
+	ddramc_config->t0pr = _at91_parameters.ddramc_t0pr;
+	ddramc_config->t1pr = _at91_parameters.ddramc_t1pr;
+	ddramc_config->t2pr = _at91_parameters.ddramc_t2pr;
 
 #else
 #error "No CLK setting defined"
 #endif
+
+	if (!ddramc_config->mdr || !ddramc_config->cr || !ddramc_config->rtr ||
+		!ddramc_config->t0pr || !ddramc_config->t1pr || !ddramc_config->t2pr)
+		dbg_info("DDRAMC: missing configuration\n");
 }
 
 static void ddramc_init(void)
diff --git a/include/common.h b/include/common.h
index 06300b8..72efa49 100644
--- a/include/common.h
+++ b/include/common.h
@@ -105,9 +105,15 @@ struct at91_param_table_t {
 	uint16_t version;
 	uint8_t boot_origin;
 	uint8_t dummy;
+
+	uint32_t ddramc_mdr;
+	uint32_t ddramc_cr;
+	uint32_t ddramc_t0pr;
+	uint32_t ddramc_t1pr;
+	uint32_t ddramc_t2pr;
 };
 
-STATIC_ASSERT(sizeof(struct at91_param_table_t) == 20, sizeof_at91_param_table);
+STATIC_ASSERT(sizeof(struct at91_param_table_t) == 40, sizeof_at91_param_table);
 
 extern const struct at91_param_table_t _at91_parameters;
 
diff --git a/main.c b/main.c
index 6abcf4b..66ec201 100644
--- a/main.c
+++ b/main.c
@@ -43,8 +43,120 @@
 #include "secure.h"
 
 
+#if defined DDRAMC_K4T51163QQ_BIE6 || \
+	defined DDRAMC_K4T1G164QG_BCE6 || \
+	defined DDRAMC_MT47H128M16RT_25E
+	#include "arch/at91_ddrsdrc.h"
+#endif
+
 const struct at91_param_table_t _at91_parameters = {
 	"AT91_PARAMETERS", 1, 0, 0,
+
+#if defined DDRAMC_K4T51163QQ_BIE6
+	/* DDRAMC MDR */
+	(AT91C_DDRC2_DBW_32_BITS
+		| AT91C_DDRC2_MD_DDR2_SDRAM),
+	/* DDRAMC CR */
+	(AT91C_DDRC2_NC_DDR10_SDR9
+		| AT91C_DDRC2_NR_13
+		| AT91C_DDRC2_CAS_5
+		| AT91C_DDRC2_DLL_RESET_DISABLED
+		| AT91C_DDRC2_DIS_DLL_DISABLED
+		| AT91C_DDRC2_NB_BANKS_4
+		| AT91C_DDRC2_DECOD_INTERLEAVED
+		| AT91C_DDRC2_UNAL_SUPPORTED),
+	/* DDRAMC TPR0 @ 200MHz */
+	(AT91C_DDRC2_TRAS_(9)
+		| AT91C_DDRC2_TRCD_(3)
+		| AT91C_DDRC2_TWR_(3)
+		| AT91C_DDRC2_TRC_(12)
+		| AT91C_DDRC2_TRP_(3)
+		| AT91C_DDRC2_TRRD_(2)
+		| AT91C_DDRC2_TWTR_(2)
+		| AT91C_DDRC2_TMRD_(2)),
+	/* DDRAMC TPR1 @ 200MHz */
+	(AT91C_DDRC2_TXP_(2)
+		| AT91C_DDRC2_TXSRD_(200)
+		| AT91C_DDRC2_TXSNR_(23)
+		| AT91C_DDRC2_TRFC_(21)),
+	/* DDRAMC TPR2 @ 200MHz */
+	(AT91C_DDRC2_TFAW_(10)
+		| AT91C_DDRC2_TRTP_(2)
+		| AT91C_DDRC2_TRPA_(3)
+		| AT91C_DDRC2_TXARDS_(3)
+		| AT91C_DDRC2_TXARD_(2)),
+
+#elif defined DDRAMC_K4T1G164QG_BCE6
+	/* DDRAMC MDR */
+	(AT91C_DDRC2_DBW_32_BITS
+		| AT91C_DDRC2_MD_DDR2_SDRAM),
+	/* DDRAMC CR */
+	(AT91C_DDRC2_NC_DDR10_SDR9
+		| AT91C_DDRC2_NR_13
+		| AT91C_DDRC2_CAS_5
+		| AT91C_DDRC2_DLL_RESET_DISABLED
+		| AT91C_DDRC2_DIS_DLL_DISABLED
+		| AT91C_DDRC2_NB_BANKS_8
+		| AT91C_DDRC2_DECOD_INTERLEAVED
+		| AT91C_DDRC2_UNAL_SUPPORTED),
+	/* DDRAMC TPR0 @ 200MHz */
+	(AT91C_DDRC2_TRAS_(9)
+		| AT91C_DDRC2_TRCD_(3)
+		| AT91C_DDRC2_TWR_(3)
+		| AT91C_DDRC2_TRC_(12)
+		| AT91C_DDRC2_TRP_(3)
+		| AT91C_DDRC2_TRRD_(2)
+		| AT91C_DDRC2_TWTR_(2)
+		| AT91C_DDRC2_TMRD_(2)),
+	/* DDRAMC TPR1 @ 200MHz */
+	(AT91C_DDRC2_TXP_(2)
+		| AT91C_DDRC2_TXSRD_(200)
+		| AT91C_DDRC2_TXSNR_(28)
+		| AT91C_DDRC2_TRFC_(26)),
+	/* DDRAMC TPR2 @ 200MHz */
+	(AT91C_DDRC2_TFAW_(10)
+		| AT91C_DDRC2_TRTP_(2)
+		| AT91C_DDRC2_TRPA_(3)
+		| AT91C_DDRC2_TXARDS_(3)
+		| AT91C_DDRC2_TXARD_(2)),
+
+#elif defined DDRAMC_MT47H128M16RT_25E
+	/* DDRAMC MDR */
+	(AT91C_DDRC2_DBW_32_BITS
+		| AT91C_DDRC2_MD_DDR2_SDRAM),
+	/* DDRAMC CR */
+	(AT91C_DDRC2_NC_DDR10_SDR9
+		| AT91C_DDRC2_NR_14
+		| AT91C_DDRC2_CAS_5
+		| AT91C_DDRC2_DLL_RESET_DISABLED
+		| AT91C_DDRC2_DIS_DLL_DISABLED
+		| AT91C_DDRC2_NB_BANKS_8
+		| AT91C_DDRC2_DECOD_INTERLEAVED
+		| AT91C_DDRC2_UNAL_SUPPORTED),
+	/* DDRAMC TPR0 @ 200MHz */
+	(AT91C_DDRC2_TRAS_(9)
+		| AT91C_DDRC2_TRCD_(3)
+		| AT91C_DDRC2_TWR_(3)
+		| AT91C_DDRC2_TRC_(12)
+		| AT91C_DDRC2_TRP_(3)
+		| AT91C_DDRC2_TRRD_(2)
+		| AT91C_DDRC2_TWTR_(2)
+		| AT91C_DDRC2_TMRD_(2)),
+	/* DDRAMC TPR1 @ 200MHz */
+	(AT91C_DDRC2_TXP_(2)
+		| AT91C_DDRC2_TXSRD_(200)
+		| AT91C_DDRC2_TXSNR_(41)
+		| AT91C_DDRC2_TRFC_(39)),
+	/* DDRAMC TPR2 @ 200MHz */
+	(AT91C_DDRC2_TFAW_(10)
+		| AT91C_DDRC2_TRTP_(3)
+		| AT91C_DDRC2_TRPA_(3)
+		| AT91C_DDRC2_TXARDS_(3)
+		| AT91C_DDRC2_TXARD_(2)),
+
+#else
+	0, 0, 0, 0, 0
+#endif
 };
 
 
-- 
1.9.1

